plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
-	p.skipFirstIteration = false
+	p.CPUMetrics, err = parseCPUMetrics(p.CPUMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to parse core metrics: %w", err)
+	}
 
-	return nil
-}
+	if len(p.CPUMetrics) == 0 && len(p.PackageMetrics) == 0 {
+		return fmt.Errorf("no metrics were found in the configuration file")
+	}
 
-func (p *PowerStat) addGlobalMetrics(acc telegraf.Accumulator) {
-	// Prepare RAPL data each gather because there is a possibility to disable rapl kernel module
-	p.rapl.initializeRaplData()
-	for socketID := range p.rapl.getRaplData() {
-		if p.packageTurboLimit {
-			p.addTurboRatioLimit(socketID, acc)
-		}
+	p.parsedCPUTimedMsrMetrics = parseCPUTimeRelatedMsrMetrics(p.CPUMetrics)
+	p.parsedCPUPerfMetrics = parseCPUPerfMetrics(p.CPUMetrics)
 
-		if p.packageUncoreFrequency {
-			die := maxDiePerSocket(socketID)
-			for actualDie := 0; actualDie < die; actualDie++ {
-				p.addUncoreFreq(socketID, strconv.Itoa(actualDie), acc)
-			}
-		}
+	p.parsedPackageRaplMetrics = parsePackageRaplMetrics(p.PackageMetrics)
+	p.parsedPackageMsrMetrics = parsePackageMsrMetrics(p.PackageMetrics)
 
-		if p.packageCPUBaseFrequency {
-			p.addCPUBaseFreq(socketID, acc)
-		}
+	if len(p.ExcludedCPUs) != 0 && len(p.IncludedCPUs) != 0 {
+		return errors.New("configuration error. Provide either one 'included_cpus'/'excluded_cpus' configuration option, or none")
+	}
 
-		err := p.rapl.retrieveAndCalculateData(socketID)
+	if len(p.ExcludedCPUs) != 0 {
+		p.parsedExcludedCores, err = parseCores(p.ExcludedCPUs)
 		if err != nil {
-			// In case of an error skip calculating metrics for this socket
-			if val := p.logOnce[socketID+"rapl"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("Error fetching rapl data for socket %s, err: %v", socketID, err)
-				// Remember that specific error occurs for socketID to omit logging next time
-				p.logOnce[socketID+"rapl"] = err
-			}
-			continue
+			return fmt.Errorf("failed to parse excluded CPUs: %w", err)
 		}
+	}
 
-		// If error stops occurring, clear logOnce indicator
-		p.logOnce[socketID+"rapl"] = nil
-		if p.packageThermalDesignPower {
-			p.addThermalDesignPowerMetric(socketID, acc)
+	if len(p.IncludedCPUs) != 0 {
+		p.parsedIncludedCores, err = parseCores(p.IncludedCPUs)
+		if err != nil {
+			return fmt.Errorf("failed to parse included CPUs: %w", err)
 		}
+	}
 
-		if p.skipFirstIteration {
-			continue
-		}
-		if p.packageCurrentPowerConsumption {
-			p.addCurrentSocketPowerConsumption(socketID, acc)
-		}
-		if p.packageCurrentDramPowerConsumption {
-			p.addCurrentDramPowerConsumption(socketID, acc)
+	p.needsCoreFreq = needsCoreFreq(p.CPUMetrics)
+	p.needsMsrCPU = needsMsr(p.CPUMetrics)
+	p.needsPerf = needsPerf(p.CPUMetrics)
+	if p.needsPerf {
+		if err = checkFile(p.EventDefinitions); err != nil {
+			return fmt.Errorf("failed to parse event definitions path: %w", err)
 		}
```

Comment:
Can you please move the code of `checkFile` in here and not split the two!? There is no reuse of that function AFAICS...

Reply:
Yeah I would like to merge the two. The main reason is that, without reuse, you have to jump around in the file when reading the code to get a full understanding. Factoring out functions is OK if they are either reused or so complex that it hampers readability, but this is not the case here IMO.

Reply:
Yes I agree, thanks for the explanation. Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).